internal/cmd/login.go

```diff
@@ -373,69 +400,113 @@ func runSignupForLogin(cli *CLI, client *api.Client) (*api.LoginRequestPasswordC
+	}
+	c := loginClient{
+		APIClient:          apiClient(options.Server, "", httpTransportForHostConfig(cfg)),
+		TrustedCertificate: options.TrustedCertificate,
+	}
+	if options.SkipTLSVerify {
+		return c, nil
+	}
 
-			if !errors.Is(err, ErrTLSNotVerified) {
-				return nil, err
-			}
+	// Prompt user only if server fails the TLS verification
+	if err := attemptTLSRequest(options); err != nil {
+		var uaErr x509.UnknownAuthorityError
+		if !errors.As(err, &uaErr) {
+			return c, err
+		}
 
-			if options.NonInteractive {
-				fmt.Fprintf(os.Stderr, "%s\n", ErrTLSNotVerified.Error())
-				return nil, Error{Message: "Non-interactive login does not allow insecure connection by default,\n      to allow, run with '--skip-tls-verify'"}
+		if options.NonInteractive {
+			// TODO: add the --tls-ca flag
+			// TODO: give a different error if the flag was set
+			return c, Error{
+				Message: "The authenticity of the server could not be verified. " +
+					"Use the --tls-ca flag to specify a trusted CA, or run " +
+					"in interactive mode.",
 			}
+		}
 
-			if err = promptSkipTLSVerify(cli); err != nil {
-				return nil, err
-			}
-			options.SkipTLSVerify = true
+		if err = promptVerifyTLSCert(cli, uaErr.Cert); err != nil {
+			return c, err
 		}
-	}
 
-	client := apiClient(options.Server, "", options.SkipTLSVerify)
-	return client, nil
+		pool, err := x509.SystemCertPool()
+		if err != nil {
+			return c, err
+		}
+		pool.AddCert(uaErr.Cert)
+		transport := &http.Transport{
+			TLSClientConfig: &tls.Config{
+				// set min version to the same as default to make gosec linter happy
+				MinVersion: tls.VersionTLS12,
+				RootCAs:    pool,
+			},
+		}
+		c.APIClient = apiClient(options.Server, "", transport)
+		c.TrustedCertificate = string(certs.PEMEncodeCertificate(uaErr.Cert.Raw))
+	}
+	return c, nil
 }
 
-func verifyTLS(host string) error {
-	url, err := urlx.Parse(host)
+func attemptTLSRequest(options loginCmdOptions) error {
+	reqURL := "https://" + options.Server
+	// First attempt with the system cert pool
+	req, err := http.NewRequestWithContext(context.TODO(), http.MethodGet, reqURL, nil)
 	if err != nil {
-		logging.S.Debug("Cannot parse host", host, err)
-		logging.S.Error("Could not login. Please check the server hostname")
-		return err
+		return fmt.Errorf("failed to create request: %w", err)
 	}
-	url.Scheme = "https"
-	urlString := url.String()
 
-	req, err := http.NewRequestWithContext(context.TODO(), http.MethodGet, urlString, nil)
-	if err != nil {
-		logging.S.Debugf("Cannot create request: %v", err)
-		return err
+	logging.S.Debugf("call server: test tls for %q", reqURL)
+	httpClient := http.Client{Timeout: 60 * time.Second}
+	res, err := httpClient.Do(req)
+	if err == nil {
+		res.Body.Close()
+		return nil
 	}
 
-	logging.S.Debugf("call server: test tls for %q", host)
-	res, err := http.DefaultClient.Do(req)
+	// Second attempt with an empty cert pool. This is necessary because at least
```

Comment:
Any open issue for this in the Go project?

Reply:
I couldn't find one, so I created https://github.com/golang/go/issues/53401.

I'll add a link to the comment.